### PR TITLE
Add SHUFFLE clause support for SELECT #2

### PR DESCRIPTION
**Pull request by xuir268**
### Changelog category (leave one):
- Experimental Feature

### Changelog entry (https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Added the experimental `LIMIT ... SHUFFLE` clause for `SELECT`, which returns a random sample of rows. Enable it with `allow_experimental_shuffle_query` and `enable_analyzer = 1`.

Documentation entry for user-facing changes
 Documentation is written (mandatory for new features)
 
...

---
Workflow [[PR](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=105338&sha=latest&name_0=PR)]
Sync PR [[sync-upstream/pr/105338](https://github.com/search?q=head%3Async-upstream%2Fpr%2F105338+org%3AClickHouse+type%3Apr&type=pullrequests)]

### CODE REVIEW

**Comment by xuir268** [comment]
@shankar-iyer @alexey-milovidov you can review the MR again

**Comment by alexey-milovidov** [comment]
The failures of `00175_obfuscator_schema_inference` and `00096_obfuscator_save_load` in `Stateless tests (amd_tsan, parallel)` are NOT caused by this PR.

They were introduced by #104690 ("Add `UntrackedMemory` asynchronous metric"), which made `clickhouse-obfuscator` abort (SIGABRT) on process teardown: the query results are correct, but the `Aborted` line on stderr fails the test. #104690 has now been reverted (#106365).

#104690 was merged in violation of the ClickHouse team rules: its own CI already showed these two tests failing (10 times between May 12 and June 1) before it was merged.

Please update your branch to pick up the revert; the tests should pass again.

**Comment by xuir268** [comment]
@shankar-iyer you can review this MR

**Comment by xuir268** [comment]
@alexey-milovidov @shankar-iyer the MR is ready to review

**Comment by xuir268** [comment]
@alexey-milovidov @shankar-iyer do we need add something more and please review it

File: src/Interpreters/Cache/QueryResultCache.cpp [hunk 1]
```diff
@@ -71,6 +83,472 @@ namespace ErrorCodes
...
+            return false;
+
+        if (nested_function->name == "view" || nested_function->name == "viewIfPermitted")
+        {
+            const auto * query = nested_function->tryGetQueryArgument();
+            return !query || astOrViewDependenciesHaveLimitShuffle(
+                query->clone(), context, visited_views, /*fail_on_missing_dependency*/ true);
+        }
+
+        return true;
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Interpreters/Cache/QueryResultCache.cpp:190]: Addressed: `isSelfContainedDeterministicRemoteTableFunction` now allowlists `numbers`, `numbers_mt`, `zeros`, `zeros_mt`, `generateSeries`, `generate_series` and `values`, and `view`/`viewIfPermitted` arguments are inspected recursively, so e.g. `SELECT * FROM remote('localhost', generateSeries(1, 3)) SETTINGS use_query_cache = 1` is cacheable again. The fail-close path remains only for wrappers that can hide catalog objects (`merge`, `loop`, nested `remote`/`cluster`, etc.). Covered by the new cases in `04257_limit_shuffle.sql`.

File: src/Interpreters/InterpreterExplainQuery.cpp [hunk 2]
```diff
@@ -663,6 +686,9 @@ QueryPipeline InterpreterExplainQuery::executeImpl()
     InterpreterSetQuery::applySettingsFromQuery(query, explain_query_context);
     query_context = std::move(explain_query_context);
 
+    if (ast.getKind() != ASTExplainQuery::ParsedAST)
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Interpreters/InterpreterExplainQuery.cpp:1003]: Fixed in 7b4895401d9: the guard in `ExplainAnalyzedSyntaxMatcher` no longer depends on the context analyzer setting — the `EXPLAIN SYNTAX` fallback for non-SELECT top-level statements force-disables the analyzer in the context, which defeated the previous check. Since `checkLimitShuffleWithAnalyzer` already rejects `EXPLAIN` of `LIMIT SHUFFLE` when the user's analyzer is disabled, the visitor now simply skips the legacy analysis (only needed for view expansion) for such selects. `EXPLAIN SYNTAX INSERT INTO ... SELECT ... LIMIT 1 SHUFFLE` works now and is covered in `04257_limit_shuffle.sql` with its output recorded in the reference.

**Comment by xuir268** [comment]
@alexey-milovidov all comments are resolved now

File: src/Interpreters/Cache/QueryResultCache.cpp [hunk 3]
```diff
@@ -71,6 +83,483 @@ namespace ErrorCodes
...
+    {
+        if (const auto * function = table_expression->table_function ? table_expression->table_function->as<ASTFunction>() : nullptr)
+        {
+            if (parameterizedViewFunctionHasLimitShuffle(*function, context, visited_views))
+                return true;
+            if (tableFunctionHasLimitShuffle(*function, context, visited_views))
+                return true;
+        }
+    }
+    else if (const auto * function = ast->as<ASTFunction>())
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Interpreters/Cache/QueryResultCache.cpp:547]: 🕵 I could not reproduce this one — nesting a parameterized view inside another table function is not supported by ClickHouse at all, independently of the query cache, so neither bad outcome is reachable.

Against this branch (26.8.1.1, `pv` = a parameterized view whose stored query uses `LIMIT SHUFFLE`):

```
SELECT * FROM loop(t5.pv(n = 10)) LIMIT 1 SETTINGS enable_analyzer = 1, use_query_cache = 1;
Code: 46. Unknown table function t5.pv. (UNKNOWN_FUNCTION)

SELECT * FROM remote('127.0.0.1', t5.pv(n = 10)) SETTINGS enable_analyzer = 1, use_query_cache = 1;
Code: 47. Unknown expression or function identifier `n` in scope `t5.pv`(n = 10). (UNKNOWN_IDENTIFIER)

SELECT * FROM cluster('test_shard_localhost', t5.pv(n = 10)) SETTINGS enable_analyzer = 1, use_query_cache = 1;
Code: 47. Unknown expression or function identifier `n` in scope `t5.pv`(n = 10). (UNKNOWN_IDENTIFIER)
```

Both errors come from analysis of the table function argument and are identical with `use_query_cache = 0`, i.e. they are pre-existing behavior of `loop` / `remote` / `cluster` over a parameterized view and not something the query result cache check introduces or hides. So `loop(pv(...))` cannot cache and replay a `LIMIT SHUFFLE` sample, and `remote` / `cluster` never reach the literal `db` / `table` branch of `remoteTableTargetHasLimitShuffle` with a parameterized view call — analysis fails first.

The reachable shapes are all already covered and asserted in `04257_limit_shuffle`: a direct `pv(n = 10)` read (handled by the `ASTTableExpression` branch), a parameterized view behind `Merge` / `Alias` / `Distributed` (handled via `storageHasLimitShuffle` → `viewStorageHasLimitShuffle`, which matches any `StorageView`), and `cluster(..., view(SELECT ... LIMIT SHUFFLE))`.

I did prototype the suggested change (calling `parameterizedViewFunctionHasLimitShuffle` for every plain `ASTFunction` node) but dropped it: with no reachable case it only adds a `DatabaseCatalog` lookup per function node of every cache-eligible query. Happy to add it if you have a query shape that actually gets that far.

File: src/Interpreters/InterpreterExplainQuery.cpp [hunk 4]
```diff
@@ -255,6 +304,27 @@ namespace
                 StorageView::replaceWithSubquery(select, query_info.view_query->clone(), tmp, query_info.is_parameterized_view);
             }
         }
+
+        static bool selectReadsFromLimitShuffleView(const ASTSelectQuery & select, const ContextPtr & context)
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Interpreters/InterpreterExplainQuery.cpp:321]: 🕵 Confirmed and fixed in f4c21114bde. Reproduced exactly as described, and probing it surfaced a second shape of the same gap, so both are now covered:

1. **The call left intact** (`FINAL` / `SAMPLE`, or a view that is not `SQL SECURITY INVOKER`) — it reaches the guard as `table_expression->table_function`, which `selectReadsFromLimitShuffleView` did not look at. It now resolves such a call as a parameterized view through a new `tryGetParameterizedViewStorage` helper (also reused by `ExpandParameterizedViewsMatcher`, so the table-function-precedence rule lives in one place) and checks the stored query.

2. **The call inlined into a subquery** by `ExpandParameterizedViewsMatcher` — this failed too, for a different reason: `needChildVisit` stops at the enclosing select, so `select.limit_shuffle` never sees the `LIMIT SHUFFLE` that was just inlined below it, and `InterpreterSelectQuery` rejected it. The check is now done over the whole select subtree with `hasLimitShuffle`.

Before, both threw `Code: 344 ... Support for LIMIT SHUFFLE requires the query analyzer` / `SUPPORT_IS_DISABLED`; now they render the syntax, matching what execution accepts (verified that the equivalent `INSERT ... SELECT * FROM pv(n = 10) FINAL` runs fine). New cases in `04257_limit_shuffle` cover both; the inlined case deliberately uses a `numbers`-backed view so the rendered subquery carries no database name and the reference stays environment-independent.

**Comment by alexey-milovidov** [comment]
🕵 On the AI review's third point (no before/after evidence for the "optimized SHUFFLE LIMIT execution" claim) — I measured it locally on this branch rather than adding a CI performance test, since the feature is off by default and the comparison is between two shapes of the *same* build.

Table: 3M rows, `MergeTree ORDER BY id`, 7 columns (3 × `String(60)`, `Array(UInt64)`, `DateTime`, `Float64`), 4 active parts, 38.8 MiB on disk. Best of 5–7 runs each, `FORMAT Null`, `enable_analyzer = 1`.

| n | `LIMIT n SHUFFLE` | `ORDER BY rand() LIMIT n` |
|---|---|---|
| 10 | 0.026s | 0.026s |
| 1000 | 0.086s | 0.094s |
| 100000 | 0.300s | 0.301s |
| 500000 | 0.472s | 0.473s |

The query plans are structurally identical (`Sorting` → `Limit` → `JoinLazyColumnsStep` + `LazilyReadFromMergeTree` in both); the only difference is the internal unordered-output flag that lets `SortingStep` and lazy materialization skip preserving the randomized order. So: no regression, and no measurable win in this shape either — the wall clock is dominated by the read and the `rand()` sort, both of which are common to the two formulations.

I have therefore not claimed a speedup in the changelog entry: I rewrote it to describe only the user-visible behavior, as the review asked, and moved the category to `Experimental Feature`. Whether the unordered-output path deserves a dedicated performance test is a judgement call for @alexey-milovidov / @shankar-iyer — I did not want to add a permanently-running perf test for an experimental, default-off feature on the strength of a null result.

File: src/Parsers/ASTSelectQuery.cpp [hunk 5]
```diff
@@ -54,6 +54,7 @@ void ASTSelectQuery::updateTreeHashImpl(SipHash & hash_state, bool ignore_aliase
     hash_state.update(group_by_with_rollup);
     hash_state.update(group_by_with_cube);
     hash_state.update(limit_with_ties);
+    hash_state.update(limit_shuffle);
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Parsers/ASTSelectQuery.cpp:73]: 🕵 Fixed in 331659af240f: `limit_shuffle` now round-trips through AST JSON and JSON deserialization rejects every parser-impossible combination.

File: src/Interpreters/Cache/QueryResultCache.cpp [hunk 6]
```diff
@@ -71,6 +83,483 @@ namespace ErrorCodes
...
+            return false;
+
+        if (nested_function->name == "view" || nested_function->name == "viewIfPermitted")
+        {
+            const auto * query = nested_function->tryGetQueryArgument();
+            return !query || astOrViewDependenciesHaveLimitShuffle(
+                query->clone(), context, visited_views, /*fail_on_missing_dependency*/ true);
+        }
+
+        return true;
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Interpreters/Cache/QueryResultCache.cpp:190]: 🕵 Fixed in 331659af240f: nested `merge`, `loop`, `cluster`, and `remote` targets now recurse through the existing dependency checks; opaque table functions remain fail-closed. `04257_limit_shuffle` covers the plain `merge` target.

**Comment by xuir268** [comment]
@alexey-milovidov @shankar-iyer I pushed 6b3659b0eb999372040fa4a9ecaefeaf0ad71787 to address the current CI flags on PR #105338.\n\nChanges included:\n- moved `allow_experimental_shuffle_query` into the current `26.9` settings-history block;\n- added `reference/statements/select/limit-shuffle` to docs navigation;\n- removed the direct generated-block edit from the `SELECT` overview by restoring that generated content;\n- refreshed `04257_limit_shuffle.reference` for the current single-record `EXPLAIN SYNTAX` output.\n\nLocal checks passed for docs navigation, generated-region validation against the CI merge base, staged whitespace, and focused compile of `SettingsChangesHistory.cpp`. Could you please check whether any more changes are required?

**Comment by xuir268** [comment]
@alexey-milovidov , is there any more changes do we need

**Comment by alexey-milovidov** [comment]
🕵 @xuir268 — I updated the branch for you (pushed `2e4631836b3`). The pull request was `CONFLICTING` and the `Fast test` / `Build (arm_tidy)` jobs were red on `d4c42fc9`; both are now addressed and the pull request is `MERGEABLE`.

**Merged `master`** (the branch was 438 commits behind). Four conflicts:

- `src/Core/SettingsChangesHistory.cpp` — kept both entries in the 26.9 block.
- `src/Interpreters/InterpreterSelectQuery.cpp` — kept both new includes.
- `src/Parsers/ParserSelectQuery.cpp` — master added the `LIMIT AFTER ... UNTIL ...` range syntax and moved the count parsing under `if (!parseLimitRange(...))`. Worth flagging: the mechanical merge placed the `SHUFFLE` branch as an `else` of that *outer* check, where it is **dead code** — `parseLimitRange` returns false for `LIMIT n SHUFFLE`, so the block is entered and the branch is never reached, and `LIMIT n SHUFFLE` would have silently parsed as a plain `LIMIT n` with no `SHUFFLE` and no error. I chained it onto the `LIMIT BY` branch *inside* the block instead, matching the pre-merge structure, and also parsed `SHUFFLE` in the post-`LIMIT BY` second `LIMIT`.
- `src/Parsers/ASTSelectQuery.cpp` — kept both the `SHUFFLE` output and master's `AFTER`/`UNTIL` output in `formatImpl`, and both sets of AST JSON deserialization guards.

Since `SHUFFLE` and `LIMIT AFTER`/`UNTIL` are mutually exclusive shapes, I rejected the combination in both places master validates: `ParserSelectQuery` (`Can not use SHUFFLE with LIMIT AFTER/UNTIL`) and the `deserializeJSON` invariant check.

**Two build failures, both mechanical merge artifacts** where master changed code immediately adjacent to this branch's insertions (`e2e045b1cf6`, `2c3e950d07c`):

- `InterpreterCreateQuery.cpp` — master added `getNumberOfTablesToCreate` exactly where `hasLimitShuffleWithDisabledSetting` ended, and the merge swallowed the latter's closing brace, so the rest of the file parsed as nested inside it (the 20 `function definition is not allowed here` errors in the `Fast test` log).
- `Settings.cpp` — master removed the `DECLARE` of `allow_experimental_codecs` (it is now `MAKE_OBSOLETE`), and the merge resurrected it next to the new `allow_experimental_shuffle_query` declaration: `redefinition of enumerator`. The same artifact also resurrected the now-orphaned `extern const SettingsBool allow_experimental_codecs;` in `InterpreterCreateQuery.cpp`, which compiled fine but failed the style check.

**Also regenerated** `04257_limit_shuffle.reference`: master changed the default of `explain_syntax_single_record` to `1`, so `EXPLAIN SYNTAX` returns one row with `\n`-escaped newlines rather than one row per line.

**Verified locally** (release build, aarch64):

- `ninja clickhouse` clean.
- `04257_limit_shuffle`, `04305_limit_shuffle_mutation_guards`, `02494_query_cache_system_tables` pass.
- All **43** of master's `LIMIT AFTER`/`UNTIL` tests pass alongside them (45/45) — this was the main risk of the parser reconciliation.
- Lazy-materialization sweep: 83/90 pass; the 7 remaining need S3/Iceberg/stateful/stateless-worker infrastructure I do not have locally.
- Hand checks: `LIMIT n SHUFFLE` returns `n` rows and genuinely randomizes across runs; `SHUFFLE` with `ORDER BY` / `OFFSET` / `WITH TIES` / `AFTER` / `UNTIL` all rejected; `LIMIT 1 BY x LIMIT 2 SHUFFLE` works; the pass-2 `loop(view ORDER BY rand() LIMIT 1)` regression and the `FINAL` + filter + `LIMIT`-without-`ORDER BY` path are both fine.
- Style check: the `cpp` check passes; the only remaining local failures are the environmental `symlinks` (my `build/` directory) and `settings_changes_history` (needs the CI `changed_files` hook).

There are no unresolved review threads. The pull request still needs a maintainer review.

**Comment by xuir268** [comment]
@alexey-milovidov Thanks for your help and I really hope we can the review as earliest possible

**Comment by xuir268** [comment]
@alexey-milovidov @shankar-iyer @novikd Could you please take another look at this PR when you have time? I have pushed the latest merge-conflict resolutions and CI fixes, and CI is running again. This PR has been open for several months; I would appreciate a maintainer review of the current design and any remaining changes needed before it can be merged. Thank you.
